src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,341 @@
+
+void deriveNativeScriptHash_handleAll(read_view_t* view)
+{
+	// Parse data
+	ctx->remainingScripts[ctx->level] = parse_u4be(view);
+	ctx->type[ctx->level] = NATIVE_COMPLEX_SCRIPT_ALL;
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE("level = %u, remaining scripts = %u", ctx->level, ctx->remainingScripts[ctx->level]);
+
+	// max possible length 26: "Confirm x scripts"
+	// where x is 2^32-1
+	char text[27];
+	explicit_bzero(text, SIZEOF(text));
+	snprintf(text, SIZEOF(text), "Confirm %d scripts", ctx->remainingScripts[ctx->level]);
+	ASSERT(strlen(text) + 1 < SIZEOF(text));
+
+	ui_displayPaginatedText(
+		// TODO: proper UI screen headers with numbered sections
+		"Script - ALL",
+		text,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+void deriveNativeScriptHash_handleAny(read_view_t* view)
+{
+	// Parse data
+	ctx->remainingScripts[ctx->level] = parse_u4be(view);
+	ctx->type[ctx->level] = NATIVE_COMPLEX_SCRIPT_ANY;
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE("level = %u, remaining scripts = %u", ctx->level, ctx->remainingScripts[ctx->level]);
+
+	// max possible length 26: "Confirm x scripts"
+	// where x is 2^32-1
+	char text[27];
+	explicit_bzero(text, SIZEOF(text));
+	snprintf(text, SIZEOF(text), "Confirm %d scripts", ctx->remainingScripts[ctx->level]);
+	ASSERT(strlen(text) + 1 < SIZEOF(text));
+
+	ui_displayPaginatedText(
+		"Script - ANY",
+		text,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+void deriveNativeScriptHash_handleNofK(read_view_t* view)
+{
+	// parse data
+	ctx->nativeScript.nOfK.n = parse_u4be(view);
+	ctx->remainingScripts[ctx->level] = parse_u4be(view);
+	ctx->type[ctx->level] = NATIVE_COMPLEX_SCRIPT_N_OF_K;
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE("level = %u, required scripts = %u, remaining scripts = %u", ctx->level, ctx->nativeScript.nOfK.n, ctx->remainingScripts[ctx->level]);
+	
+	// validate that the received requiredScripts count makes sense
+	VALIDATE(ctx->remainingScripts[ctx->level] >= ctx->nativeScript.nOfK.n, ERR_INVALID_DATA);
+
+	// max possible length 32: "Confirm x of x"
+	// where x is 2^32-1
+	char text[33];
+	explicit_bzero(text, SIZEOF(text));
+	snprintf(text, SIZEOF(text), "Confirm %d of %d", ctx->nativeScript.nOfK.n, ctx->remainingScripts[ctx->level]);
+	ASSERT(strlen(text) + 1 < SIZEOF(text));
+
+	ui_displayPaginatedText(
+		"Script - N of K",
+		text,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+void deriveNativeScriptHash_handleComplexScriptStart(read_view_t* view)
+{
+	TRACE_BUFFER(view->begin, (size_t) (view->end - view->begin));
+
+	VALIDATE(expectsRequest(), ERR_INVALID_STATE);
+
+	// check if we can increase the level without breaking the MAX_SCRIPT_DEPTH constraint
+	VALIDATE(ctx->level + 1 < MAX_SCRIPT_DEPTH, ERR_DATA_TOO_LARGE);
+	ctx->level++;
+
+	uint8_t nativeScriptType = parse_u1be(view);
+	TRACE("native complex script type = %u", nativeScriptType);
+
+	switch(nativeScriptType) {
```

Comment:
Does this conform to our astyle rules? Try `make format`.

Reply:
Added fix commits which apply the formatting as defined by `make format`.

Reply:
Now everything seems to be okay according to `make format`